book/src/error_handling.md

```diff
@@ -0,0 +1,173 @@
+   those related to Exception and resource release. Refer to the spec for the
+   full list.
+   
+## Java Exceptions and `JNIEnv`
+`JNIEnv` is the main interface that native methods use to interact with Java.
+Since they can all fail, they all return `Result` types. Refer to the [docs on
+`JNIEnv`](https://docs.rs/jni/0.18.0/jni/struct.JNIEnv.html) to understand how
+it handles errors. Note that the methods on `JNIEnv` never clear Exceptions.
+
+## Throwing Exceptions from Rust
+
+Java can't handle Rust `Result` or `Option` types, but it can handle exceptions.
+The Rust side can use
+[`throw_new`](https://docs.rs/jni/0.18.0/jni/struct.JNIEnv.html#method.throw_new)
+to throw exceptions, so that Java can decide on how to recover. For example:
+
+```rust
+#[no_mangle]
+pub extern "system" fn Java_jni_1rs_1book_NativeAPI_divide(
+    env: JNIEnv,
+    _class: JClass,
+    numerator: jint,
+    divisor: jint,
+) -> jint {
+    if let Some(result) = numerator.checked_div(divisor) {
+        // `checked_div` is a special version of division that returns `None` if the
+        // divisor is zero. Since it returned `Some`, we know it is not.
+        result
+    } else {
+        // Divisor is zero -- throw an exception!
+        env.throw_new("java/lang/ArithmeticException", "Attempting to divide by zero.");
+        // A dummy value must be returned, since the native method retains control
+        // even after throwing an Exception.
+        0
+    }
+}
+```
+
+
+    
+## Wrapping Result Handling into a Function
+
+It can be helpful to wrap common error handling into one function, so that there
+is less repeated JNI code. For ease of implementation, we'll use
+`anyhow::Error`, and make the caller pass the `dummy_value`, which is returned
+only on error paths. The `JNIEnv` APIs all throw `Result<T,E>`, and the
+`JavaException` variant marks that an exception has already been thrown - so for
+that variant alone, we will skip throwing an Exception. You may come up with
+your own Error hierarchies, which could have a different logic.
+
+```rust
+fn try_java<F, T>(env: JNIEnv, dummy_value: T, f: F) -> T
+where
+    F: FnOnce() -> Result<T, anyhow::Error>,
+{
+    match f() {
+        Ok(s) => s,
+        Err(e) => {
+            if let Some(e) = e.downcast_ref::<jni::errors::Error>() {
+                match e {
+                    // Since JavaException implies an Exception has already been thrown, 
+                    // don't throw another one.
+                    jni::errors::Error::JavaException => {},
+                    _ => {
+                        let _ = env.throw_new("java/lang/RuntimeException", e.to_string());
+                    }
+                }
+            } else {
+            // This branch handles all other error types, such as those returned by
+            // all non-JNI code
+                let _ = env.throw_new("java/lang/RuntimeException", e.to_string());
+            }
+            default_value
```

Comment:
Found this MR while looking for information on error handling with this crate.

Found this mistake where you call a value `default_value` whilst the parameter is actually `dummy_value` so either

```suggestion
            dummy_value
```

Reply:
Thanks, there are a few typos in the book. I'm going to update this tomorrow with a set of changes that includes the fix. (I switched the code samples to use Anchor tags, so that we can catch these syntactic errors more easily in the future).

Thank you for commenting that you found this though, I'd appreciate you examining the diff once it's out in case I made other mistakes.